subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java

```diff
@@ -16,101 +16,138 @@
+        this.resolveStateFactory = resolveStateFactory;
+
+        this.components = cacheFactory.createCalculatedValueCache(Describables.of("project components"), this::loadOrCreateLocalComponentState);
     }
 
     @Override
-    public LocalComponentGraphResolveState getComponent(ProjectComponentIdentifier projectIdentifier, Path currentBuildPath) {
-        boolean isLocalProject = projectIdentifier.getBuild().getBuildPath().equals(currentBuildPath.getPath());
-        if (isLocalProject) {
-            return originalComponents.get(projectIdentifier);
-        } else {
-            return foreignIdentifiedComponents.get(projectIdentifier);
-        }
+    public LocalComponentGraphResolveState getComponent(ProjectIdentity targetProject, Path sourceBuild) {
+        Path projectIdentityPath = targetProject.getBuildTreePath();
+        ensureBuildConfigured(targetProject.getBuildPath(), sourceBuild);
+        return components.get(projectIdentityPath);
     }
 
-    private LocalComponentGraphResolveState createLocalComponent(ProjectComponentIdentifier projectIdentifier) {
-        return localComponentCache.computeIfAbsent(
-            ((DefaultProjectComponentIdentifier) projectIdentifier).getIdentityPath(),
-            path -> localComponentProvider.getComponent(projectStateRegistry.stateFor(path))
-        );
+    private LocalComponentGraphResolveState loadOrCreateLocalComponentState(Path projectIdentityPath) {
+        return localComponentCache.computeIfAbsent(projectIdentityPath, path -> {
+            ProjectState projectState = projectStateRegistry.stateFor(path);
+
+            // The project may not have been configured yet if configure-on-demand is enabled,
+            // or if a parent project resolves a dependency on this project during configuration-time.
+            projectState.ensureConfigured();
+
+            return projectState.fromMutableState(p -> createLocalComponentState(projectState, p));
+        });
     }
 
-    /**
-     * Return a copy of the component identified by {@code projectIdentifier}, except with its identifier replaced with the foreign counterpart.
-     *
-     * <p>Eventually, in Gradle 9.0, when {@link BuildIdentifier#isCurrentBuild()} is removed, all this logic can disappear.</p>
-     */
-    private LocalComponentGraphResolveState copyComponentWithForeignId(ProjectComponentIdentifier projectIdentifier) {
-        ProjectState projectState = projectStateRegistry.stateFor(projectIdentifier);
-        CompositeBuildParticipantBuildState buildState = (CompositeBuildParticipantBuildState) projectState.getOwner();
-        if (buildState instanceof IncludedBuildState) {
-            // Make sure the build is configured now (not do this for the root build, as we are already configuring it right now)
-            buildState.ensureProjectsConfigured();
-        }
+    private LocalComponentGraphResolveState createLocalComponentState(ProjectState projectState, ProjectInternal project) {
```

Comment:
🤔 This should only need a `project` param, right?  You can get the state from the project directly.

Reply:
Removed the param